cli_tools/osconfig_agent/patch_window.go

```diff
@@ -15,26 +15,306 @@
+}
+
+// update updates a patchWindow with a PatchPolicy.
+func (w *patchWindow) update(p *osconfigpb.LookupConfigsResponse_EffectivePatchPolicy) {
+	fmt.Println("DEBUG: update", w.Name)
+	start, end, err := nextWindow(time.Now().UTC(), p.PatchWindow, 0)
+	if err != nil {
+		return
+	}
+
+	if start != w.Start {
+		fmt.Println("DEBUG: patchWindow start changed, need to reregister:", w.Name)
+		w.t.Stop()
+		// Cancel any potential current run if we are no longer in a patch window.
+		if w.in() {
+			fmt.Println("DEBUG: patchWindow running, need to cancel:", w.Name)
+			close(w.cancel)
+		}
+
+		defer w.register()
+	}
+
+	w.mx.Lock()
+	w.Start, w.End = start, end
+	w.Policy = &patchPolicy{p}
+	w.mx.Unlock()
+
+	fmt.Println("DEBUG: update end", w.Name)
+}
+
+// register registers a patch window as active, this will clobber any existing
+// window with the same name.
+func (w *patchWindow) register() {
+	fmt.Println("DEBUG: register", w.Name)
+	w.mx.Lock()
+	defer w.mx.Unlock()
+	fmt.Println("DEBUG: register got lock", w.Name)
+	defer func() { fmt.Println("DEBUG: register unlock", w.Name) }()
+
+	aw.add(w.Name, w)
+
+	// Create the Timer that will kick off the patch process.
+	// If we happen to be in the patch window now this will start imediately.
+	w.cancel = make(chan struct{})
+	w.t = time.AfterFunc(w.Start.Sub(time.Now()), func() { rc <- w })
+	fmt.Println(w.Name, "register done")
+}
+
+// deregister stops a patch windows timer and removes it from activeWindows
+// but does not cancel any current runs.
+func (w *patchWindow) deregister() {
+	fmt.Println(w.Name, "deregister")
+	w.t.Stop()
+	aw.delete(w.Name)
+}
+
+func (w *patchWindow) run() (reboot bool) {
+	fmt.Println("DEBUG: run", w.Name)
+
+	w.mx.RLock()
+	defer w.mx.RUnlock()
+	fmt.Println("DEBUG: run got lock", w.Name)
+	defer func() { fmt.Println("DEBUG: run unlock", w.Name) }()
+
+	defer func() {
+		if reboot {
+			if err := saveState(state, w); err != nil {
+				log.Println("ERROR:", err)
+			}
+		} else {
+			if err := saveState(state, nil); err != nil {
+				log.Println("ERROR:", err)
+			}
+		}
+	}()
+
+	// Make sure we are still in the patch window.
+	if !w.in() {
+		fmt.Println(w.Name, "not in patch window")
+		return false
+	}
+
+	fmt.Println("running patch window", w.Name)
+	if err := saveState(state, w); err != nil {
+		log.Println("ERROR:", err)
+	}
+
+	//runUpdates()
+	// Pretend to do work
+	time.Sleep(5 * time.Second)
+
+	// Make sure we are still in the patch window after each step.
+	// TODO: Pass this to runUpdates instead?
+	if !w.in() {
+		fmt.Println(w.Name, "timedout")
+		return false
+	}
+
+	fmt.Println("finished patch window", w.Name)
+	return false
+}
+
+func patchManager(efps []*osconfigpb.LookupConfigsResponse_EffectivePatchPolicy) {
+	// Deregister any existing patchWindows that no longer exist.
+	var ppns []string
+	for _, pp := range efps {
+		ppns = append(ppns, pp.GetFullName())
+	}
+
+	var toDeregister []*patchWindow
+	aw.mx.Lock()
```

Comment:
shouldn't the Unlock() be deferred here in case there's a panic or the code is refactored to return early?

Reply:
Usually yes, a panic exist so that's not an issue but early returns are. I specifically don't defer here as I need to unlock before function end, and cars is taken to not have anything return early before then.